keps/sig-auth/1688-dynamic-authentication-config/README.md

```diff
@@ -0,0 +1,620 @@
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high quality user-focused
+documentation such as release notes or a development roadmap.  It should be
+possible to collect this information before implementation begins in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself.  KEP editors, SIG Docs, and SIG PM
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+-->
+
+Authentication is Kubernetes is quite flexible and unopinionated.  There is no
+requirement for the API server to understand where an identity originates from.
+The API server supports a variety of command line flags that enable authentication
+via x509 certificates, OpenID Connect ID tokens, arbitrary bearer tokens via a
+token webhook, etc.  Dynamic authentication config expands this capability to a
+Kubernetes REST API.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Configuring authentication via command line flags has some limitations.  Some
+forms of config are simply easier to specify and understand in a structured
+manner such as via a Kubernetes API or file.  The existing authentication flags
+for OIDC and token webhook limit the user to only one of each of these types of
+authentication modes as there is no way to specify the set of flags multiple
+times [#71162].  These flag based configs also require an API server restart to
+take effect.
+
+Since there is no way to configure authentication other than via flags, hosted
+environments have no consistent way to expose this functionality [#166].  To get
+around this limitation, users install authentication proxies that run as
+cluster-admin and impersonate the desired user [kube-oidc-proxy] [teleport].
+Other than being a gross abuse of the impersonation API, this opens the API
+server to escalation bugs caused by the proxy (such as improper handling of
+incoming request headers).  This proxy also intercepts network traffic which
```

Comment:
> this opens the API server to escalation bugs caused by the proxy (such as improper handling of incoming request headers)

Isn't this true with any authentication handler? I suppose the proposed `Prefix` requirement mitigates it to some degree.

Reply:
The concern I was raising here is that with an impersonation proxy your network diagram looks like:

```
[ user agent ] -- user creds over TLS --> [ impersonation reverse proxy ]  -- proxy credentials over TLS with impersonation headers --> [ API server ]
```

The reverse proxy has to handle the incoming request, authenticate it, sanitize it (ex: fail if incoming request has impersonation headers), and then pass it through to the API server with its own credentials.  What happens if Kube ever adds new headers that have special meaning?  Some headers may be benign from a security perspective and the proxy should pass them through.  Others could be like impersonation and change the security parameters of the request.  Having a proxy that is `cluster-admin` act a confused deputy is something I would like to avoid.

This network is far easier to protect IMO:

```
[ user agent ] -- user creds over TLS --> [ API server ]  -- possible webhook over TLS with user creds --> [ webhook ]
```